src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,95 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
 
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] storage;
+    private int size;
+
+    public ArrayList() {
+        storage = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == storage.length) {
+            grow();
+        }
+        storage[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
+        }
+        System.arraycopy(storage, index, storage, index + 1, size - index);
+        storage[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexValidation(index);
+        return storage[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        indexValidation(index);
+        storage[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexValidation(index);
+        final T oldValue = storage[index];
+        System.arraycopy(storage, index + 1, storage, index, size - index - 1);
+        storage[size - 1] = null;
+        size--;
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (storage[i] == element || storage[i] != null && storage[i].equals(element)) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Element was not found");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void grow() {
+        int newCapacity = storage.length + storage.length / 2;
+        T[] newStorage = (T[]) new Object[newCapacity];
+        System.arraycopy(storage, 0, newStorage, 0, size);
+        storage = newStorage;
+    }
+
+    private void indexValidation(int index) {
```

Comment:
```suggestion
    private void validateIndex(int index) {
```
we should use verbs for method names